tooling/SecretManager.VisualStudio/ProjectLocalSecretsManager.cs

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.ProjectSystem.Properties;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Threading;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.Extensions.SecretManager.VisualStudio
+{
+    public class ProjectLocalSecretsManager : IVsProjectSecrets, SVsProjectLocalSecrets
+    {
+        private const string UserSecretsPropertyName = "UserSecretsId";
+
+        private readonly IProjectPropertiesProvider _propertiesProvider;
+
+        public ProjectLocalSecretsManager(IProjectPropertiesProvider propertiesProvider)
+        {
+            _propertiesProvider = propertiesProvider ?? throw new ArgumentNullException(nameof(propertiesProvider));
+        }
+
+        public string SanitizeName(string name) => name;
+
+        public IReadOnlyCollection<char> GetInvalidCharactersFrom(string name) => Array.Empty<char>();
+
+        public async Task AddSecretAsync(string name, string value, CancellationToken cancellationToken = default)
+        {
+            EnsureKeyNameIsValid(name);
+            var store = await GetStoreAsync();
+            if (store.ContainsKey(name))
+            {
+                throw new ArgumentException("A secret with this name already exists.", nameof(name));
+            }
+
+            store.Set(name, value);
+            cancellationToken.ThrowIfCancellationRequested();
+            store.Save();
+        }
+
+        public async Task SetSecretAsync(string name, string value, CancellationToken cancellationToken = default)
+        {
+            EnsureKeyNameIsValid(name);
+
+            var store = await GetStoreAsync();
+
+            store.Set(name, value);
+            cancellationToken.ThrowIfCancellationRequested();
+            store.Save();
+        }
+
+        public async Task<string> GetSecretAsync(string name, CancellationToken cancellationToken = default)
+        {
+            EnsureKeyNameIsValid(name);
+            var store = await GetStoreAsync();
+            return store[name];
+        }
+
+        public async Task<IReadOnlyCollection<string>> GetSecretNamesAsync(CancellationToken cancellationToken = default)
```

Comment:
@AArnott - is there any expectation or provision for something like transactions on secret? 

For instance if a dialog is showing in VS that's been populated using one of the APIs like `GetSecretNamesAsync` and the file is updated using the commandline tools, what should happen? 

Are we assuming that those kinds of things are out of scope or not important?

Reply:
Well, we do expect the returned collection to be frozen -- not a live one that keeps changing. That is somewhat a given since the interface doc says it must be thread safe, and a collection that changes is not thread-safe to return.

But the idea of knowing when to refresh is an interesting question I hadn't thought about when I designed the interface. I think we should just table it for now though, since it would undoubtedly be costly for you to implement and we don't yet have a scenario for it -- we might eventually, but we can add another interface when and if the time comes.